### PR TITLE
[WIP] Feat/lang refmt

### LINKED ISSUE #18: Shortcut to auto refmt code
Thinking about Ctrl+Shift+I (like VSCode)

### LINKED ISSUE #137: Auto-convert RE to ML to RE when changing languages
[Reason Tools](https://github.com/reasonml/reason-tools) provides an in-browser converter from RE to ML syntax and back. Wondering if this could be adapted to Sketch so that when you switch from RE to ML, the code is automatically translated to the opposite syntax. 

For the education use-case that Sketch is so great for, this would be incredibly useful for highlighting the differences between Reason and Ocaml and to help promote cross language proficiency between the two.

### PR DESCRIPTION
**Pull request by no-stack-dub-sack**
#137 progress. This is just a WIP to show you where I'm currently at. Merge conflicts will need to be resolved at some point. 

Issue I'm having right now is isolating the first edit after lang change so we know when to trash blocksCopy, and when to pull data from that when switching back and fort without making edits. I've left you an IM about this. 

If successful, this would also close #18 regarding shortcut for pretty printing RE code

close #137

### CODE REVIEW

**Comment by thangngoc89** [comment]
I wonder why circle doesn't run for you

File: client/src_editor/Editor_Blocks.re [hunk 1]
```diff
@@ -84,8 +91,12 @@ let make =
       _children,
     ) => {
   ...component,
+  retainedProps: {
+    lang: lang,
```
Comment (thangngoc89) [inline client/src_editor/Editor_Blocks.re]: retainedProps is deprecated. You need to copy props to state instead

File: client/src_editor/Editor_Blocks_Refmt.re [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      Editor_Worker.mlToRe(. code)
+      |> then_(
+           (
+             result: Belt.Result.t(string, Worker_Evaluator.Types.Refmt.error),
+           ) => {
+           result
+           |> (
+             fun
+             | Ok(re) => callback(re, blockId)
+             | Error(error) => Js.log(error)
```
Comment (thangngoc89) [inline client/src_editor/Editor_Blocks_Refmt.re]: I added a toast in master. Could you please rebase and add it? Notify.error(a string)

File: client/src_editor/Editor_Blocks.re [hunk 3]
```diff
@@ -139,18 +166,134 @@ let make =
...
         | Block_DeleteQueued(_)
         | Block_Restore(_)
-        | Block_CaptureQueuedMeta(_, _, _)
-        | Block_UpdateValue(_, _, _) => onUpdate(newSelf.state.blocks)
+        | Block_MapRefmtToBlocks(_, _, _) => onUpdate(newSelf.state.blocks)
+        | Block_UpdateValue(blockId, newValue, diff) =>
+          /* switch(newSelf.state.blocksCopy) {
+               | None => ()
+               | Some(_) =>
+                 if (diff.origin == "+input") {
```
Comment (thangngoc89) [inline client/src_editor/Editor_Blocks.re]: It's diff -> CodeMirror.Diff.originGet 
but could you make this a black list insead of a white list? (only filter out unneeded event)

File: client/src/Shortcut.re [hunk 4]
```diff
@@ -53,10 +53,14 @@ module Store = {
...
   };
 
+  /* single source of truth for this key */
+  let refmtShortcutKey = "ctrl+shift+i";
+
   let subscribe: subscribeFun =
     (~global=false, key, handler) => {
-      if ((m^)->ShortcutMap.has(key)) {
-        warn("Overiding " ++ key ++ " shortcut");
+      /* Don't warn when overriding refmt shorcut since re-registering on each lang update */
```
Comment (thangngoc89) [inline client/src/Shortcut.re]: I'm not sure why is this needed? I don't see you registering this anywhere

Comment (no-stack-dub-sack) [reply client/src/Shortcut.re]: You're right, this is no longer needed. Forgot to remove it.

File: client/src_editor/Editor_Blocks_Refmt.re (hunk 2 above)
Comment (no-stack-dub-sack) [reply client/src_editor/Editor_Blocks_Refmt.re]: sounds good, was meaning to ask you how you'd like to handle errors like these

File: client/src_editor/Editor_Blocks.re (hunk 1 above)
Comment (no-stack-dub-sack) [reply client/src_editor/Editor_Blocks.re]: Ah, interesting. Ok.

File: client/src_editor/Editor_Blocks.re (hunk 3 above)
Comment (no-stack-dub-sack) [reply client/src_editor/Editor_Blocks.re]: I'm not sure what you mean... do you mean use a catch all instead of listing out the actions we're not reacting to? e.g. `_ => ()`

Comment (no-stack-dub-sack) [reply client/src_editor/Editor_Blocks.re]: I'm not sure what you mean... do you mean use a catch all instead of listing out the actions we're not reacting to? e.g. `_ => ()`

Comment (thangngoc89) [reply client/src_editor/Editor_Blocks.re]: I think we can filter out setValue origin and not triggering onUpdate event

**Comment by no-stack-dub-sack** [comment]
@thangngoc89 just updated the PR to remove retainedProps... this is getting closer, but the conditions are still not quite right, either for triggering the refmtAsLang action or for swapping the blocks until a change is made... I think I'm getting something wrong re: component lifecycle

I've tried various approaches, and they all seem to be failing after switching back and forth a few times, so I believe I just have some conditions set wrong somewhere. 

The way it is now in willReceiveProps looks to me like it should just swap copies back and forth until an edit is made and clears blocksCopy, but it's not quite working that way. Excuse all the logs everywhere, they will be removed once this is ready.

**Comment by thangngoc89** [comment]
No worries about the logs. I do that all the time. Maybe I should write about my trick for using logs too. Let's me clone the branch and try to understand the problem you're facing

File: client/src_editor/Editor_Blocks.re [hunk 5]
```diff
@@ -93,7 +101,18 @@ let make =
...
     initialState: makeInitialState,
     willReceiveProps: ({state}) =>
-      if (blocks != state.blocks) {
+      if (state.lang != lang) {
+        Js.log2("state.blocks", state.blocks[1]);
+        Js.log2("blocksCopy", state.blocks[1]);
+        switch (state.blocksCopy) {
+        | None => makeInitialState()
+        | Some(blocksCopy) => {
+            ...makeInitialState(),
```
Comment (thangngoc89) [inline client/src_editor/Editor_Blocks.re]: should be `...state`

**Comment by thangngoc89** [comment]
Some css touches 
image

(Change background color to `var(--c-text-sub2)`)

**Comment by no-stack-dub-sack** [comment]
@thangngoc89 I've used tapLog quite a bit throughout my work on the project as well, had a similar utility in my first reason project so was happy to see it here... very handy. Used it for logging actions as one use case: `switch(action->tapLog) {...}`

**Comment by no-stack-dub-sack** [comment]
> You can merge/cherry pick back to your branch if you like.

Ok, great, let me take a look. Thanks, and no prob! fun stuff, as always.

**Comment by no-stack-dub-sack** [comment]
hmmm.... ok, this is the general feature, but it does not have the optimizations you suggested, I was unable to get a working solution for refmtMany, but I'm open to a little guidance on this. 

The only other thing currently wrong with this approach (that I see) is that Execute is no longer fired on onUpdate in Editor_Note to prevent executing on code not yet updated to target lang, I have it instead firing when last MapRefmtToBlocks fires, so blocks are not currently being executed when switching back to blocksCopy. Need to find a way to fire execute after the state update in willRecieveProps if that's possible?

**Comment by thangngoc89** [comment]
You can't fire an event inside willReceiveProps unfortunately. but you can do that in didUpdate. And the reason for doing everything in 1 go is not because of perf and optimization. Doing what I suggested would make everything eaiser to reason about. Like receive refmt code -> trigger an execution. and not having to worry about orders or anything, (the way you're currently do it, it might leads to race conditions ). I'll write something for you

**Comment by thangngoc89** [comment]
@no-stack-dub-sack I pushed a new commit here https://github.com/Sketch-sh/sketch-sh/commit/7202dd3ce73671a7ed087a8f85387c85dff196aa for a unified interface.

The interface looks like:

```
let reToMl: (. list((id, code))) => Js.Promise.t(Belt.Result.t(list(id, code), error))
```

**Comment by thangngoc89** [comment]
You may need to trim refmt result because it always add trailing new lines

**Comment by no-stack-dub-sack** [comment]
@thangngoc89 ok, how's this implementation instead? 

A couple of notes and things left to do. 

TODO:
- still need warning/notification about how refmt lang conversion may affect large code blocks. Do you just want a modal the first time the lang switch, and show only once, set local storage flag?

Questions/Issues:
- how to notify on error when refmt logic is on worker thread? I tried passing Notify.error as a callback, but threw a DOM exception. 
- this attempts to execute code when switching langs and going back to blocksCopy state (before was only executing on refmt when changing langs). It seems to work.... most of the time. Sometimes, though, the widgets don't show up in the UI even though debugging showed that widgets still get created and are in state... not sure what's happening here??

Otherwise, once these points are addressed, this should implement everything as discussed.

**Comment by thangngoc89** [comment]
@no-stack-dub-sack I don't think we need a modal for confirmation because we preserve the last state so user can just switch back.

for notify, you need to send back an error (Belt.Result.Error) and the caller need to act on that

File: client/src_analyzer/Worker_Refmt.re [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  let logRefmtError = (code, error, targetLang) => {
+    let fromLang = ref("RE");
+    let toLang = ref("ML");
+
+    switch (targetLang) {
+    | ML => ()
+    | RE =>
+      fromLang := "ML";
+      toLang := "RE";
+    };
```
[the remark this answers is not in the document]
Comment (no-stack-dub-sack) [reply client/src_analyzer/Worker_Refmt.re]: This mutation is just on the string to give the right warning to the user when their code is unable to be reformatted, not to select the correct parser/printer, but I could still apply this logic. 

Also, currently, pretty print does not fire at all when on a ML sketch. But I can change it to warn the user instead the ML is not supported.

File: client/src_editor/Editor_Blocks.re [hunk 7]
```diff
@@ -238,22 +366,7 @@ let make =
...
-                    switch (b_data) {
-                    | B_Text(_) => acc
-                    | B_Code({bc_value}) => [(b_id, bc_value), ...acc]
-                    }
-                  )
-              )
-            )
-          ->Belt.List.reverse;
-
+        let allCodeToExecute = state.blocks->codeBlockDataPairs;
```
[the remark this answers is not in the document]
Comment (no-stack-dub-sack) [reply client/src_editor/Editor_Blocks.re]: 🎉 finally removing something instead of adding!

File: client/src_editor/Editor_Blocks.re [hunk 8]
```diff
@@ -141,14 +159,31 @@ let make =
               self.send(Block_Execute(true, BTyp_Text));
             },
           );
+        let unReg4 =
+          registerShortcut(
+            ~global=true,
+            "ctrl+shift+i",
```
[the remark this answers is not in the document]
Comment (no-stack-dub-sack) [reply client/src_editor/Editor_Blocks.re]: :+1:

File: client/src_analyzer/Worker_Analyze.re [hunk 9]
```diff
@@ -105,4 +106,65 @@ module Make = (ESig: Worker_Evaluator.EvaluatorSig) => {
...
+    | RE =>
+      fromLang := "ML";
+      toLang := "RE";
+    };
+
+    let message = {j|There was a problem reformating your $fromLang code to $toLang|j};
+    /* Notify.error(message ++ ". Check the console for details."); */
+    Js.log(message ++ "\n");
+    Js.log2("Code:\n", code);
+    Js.log2("Error:\n", error);
```
Comment (no-stack-dub-sack) [inline client/src_analyzer/Worker_Analyze.re]: currently showing the error only in the console, along with the offending code, here (and saying to check the console for more details), since I thought showing the error in the notification might be too much.

Do you want the error shown in the error toast as well?

Comment (thangngoc89) [reply client/src_analyzer/Worker_Analyze.re]: I think let's keep it there. Ideally, you would show the syntax error in the editor but let's leave it there for a moment :D

Comment (no-stack-dub-sack) [reply client/src_analyzer/Worker_Analyze.re]: 👍

File: client/src_analyzer/Worker_Refmt.re (hunk 6 above)
[the remark this answers is not in the document]
Comment (thangngoc89) [reply client/src_analyzer/Worker_Refmt.re]: Alright. sorry for not looking at this closely. Looks good to me :+1:

File: client/src_analyzer/Worker_Analyze.re (hunk 9 above)
Comment (no-stack-dub-sack) [reply client/src_analyzer/Worker_Analyze.re]: whoops, I also just realized that I pulled all this code out into Worker_Refmt and never deleted it in Worker_Analyze. Will fix that in next commit.

**Comment by no-stack-dub-sack** [comment]
@thangngoc89 ok, this should be everything, hopefully the last commit! However, the weird widgets issue still needs to be investigated. I'm kind of at a loss. Might want to take a look before deploying to see what's up?

Also, I made Notify.info able to be sticky, however, wondering if we want to add Notify.warn. If so, I thought #ffa40f could be a nice color.

image

**Comment by thangngoc89** [comment]
@no-stack-dub-sack wow. Thank you for the awesome work. I agree that a warn notify is nicer. I'll add it while investigating the weird bug
